_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+  
+     res.sendStatus(204);
+   });
+```
+##### Test deleting a user 
+Test out the delete function by making a DELETE request with an empty body to `http://localhost:3333/scim/v2/Users/3`.
+
+What response do you expect? You should get an HTTP 204 with no body in response.
+
+
+##### Test soft deleting/"deprovisioning" a user
+
+Okta implements a [soft delete aka deprovision](https://developer.okta.com/docs/reference/scim/scim-20/#delete-users) via PUT or PATCH. This will cover the scenarios such as users reassignment from services and potentially auditing purposes. Add the following code to `scim.ts` to support what the spec calls a partial delete:
+
+```
+// Soft Delete Users
+// PATCH: /Users/:userId
+// RFC Notes on Partial Update: https://www.rfc-editor.org/rfc/rfc7644#section-3.5.2 
+// Note: this does not a true "delete"; this will update the active flag to false (this is an Okta best practice)
+scimRoute.patch('/Users/:userId', passport.authenticate('bearer'), async (req, res) => {
+    // console.log('PATCH: /Users/:userId'); 
+    console.log(req.body);
+  
+     const id  = parseInt(req.params.userId);
+     const active = req.body["Operations"][0]["value"]["active"]
+     await prisma.user.update({
+       data: {
+        active
+       },
+       where: { id }
+     });
+  
+     res.sendStatus(204);
+   });   
+```
+
+Try sending a PATCH request with an empty body to `http://localhost:3333/scim/v2/Users/2`.
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:api:messages:2.0:PatchOp"
+    ],
+    "Operations": [
+        {
+            "op": "replace",
+            "value": {
+                "active": false
+            }
+        }
+    ]
+}
+```
+
+Just like the delete request, you should expect to receive a 204 response; this is in line with [Okta's docs](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-patch)
+
+## Connecting with an Identity provider
+
+Now that we have a working SCIM application, let's test it with an Identity Provider! In this workshop, we're using Okta. 
+
+We'll need to set up a few things for Okta to authenticate and interact with our local server. 
+
+### 1) Give your app a public URL
+One way to give your app a public URL or IP would be to host it on a cloud instance with DNS that you control. But if that isn't an option, you can use Localtunnel to temporarily issue a public web address to your own computer. 
+For this example, we'll use [Localtunnel](https://theboroer.github.io/localtunnel-www/). To run the tunnel, you'll start the api with the following command:
+
+```
+npm run serve-api
+```
+
+In another terminal, start the tunnel using:
+
+```
+npx localtunnel --port 3333
+```
+
+It will ask for permission to install the npm package; say yes. And when you start localtunnel, it will print a URL to your terminal, such as http:/unique-url-for-squirrels. You will need that URL when we create a SCIM app in Okta so that Okta can send requests to your Todo app.
```

Comment:
```suggestion
It will ask for permission to install the npm package; say yes. And when you start localtunnel, it will print a URL to your terminal, such as http://unique-url-for-squirrels. You will need that URL when we create a SCIM app in Okta so that Okta can send requests to your Todo app.
```